sparta/sparta/resources/Pipeline.hpp

```diff
@@ -218,24 +226,33 @@ namespace sparta
         const_iterator cend() const { return const_iterator(this, capacity()); }
 
         /*!
-         * \brief Construct a Pipeline object
+         * \brief Construct a Pipeline object by existed event set
          *
+         * \param es The pointer to existed event set
          * \param name The name of Pipeline object
          * \param num_stages The number of pipeline stages
          * \param clk The clock this pipeline synchronized to
          */
-        Pipeline(const std::string & name,
+        Pipeline(EventSet * es,
+                 const std::string & name,
                  const uint32_t num_stages,
                  const Clock * clk) :
             name_(name),
             clock_(clk),
             pipe_(name, num_stages, clk),
             event_list_at_stage_(num_stages),
             event_matrix_at_stage_(num_stages),
+            es_((es == nullptr) ? &dummy_es_ : es),
             ev_pipeline_update_
-                (&es_, name + "_update_event", CREATE_SPARTA_HANDLER(Pipeline, internalUpdate_), 1),
+                (es_, name + "_update_event", CREATE_SPARTA_HANDLER(Pipeline, internalUpdate_), 1),
             num_stages_(num_stages)
         {
+            // Only support the following Event types:
+            // 1. PhasedUniqueEvent
+            // 2. PhasedPayloadEvent<DataT>
+            if constexpr (!std::is_same_v<EventT, PhasedUniqueEvent> && !std::is_same_v<EventT, PhasedPayloadEvent<DataT>>) {
+                sparta_assert(false, "Pipeline is templated a unsupported Event type");
```

Comment:
You can make this a `static_assert` actually.

Reply:
Nice!